docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+## Efficient IO
+
+As part of the high performance work, we need new abstractions for doing efficient reading and writing to and from the network. The pattern we have settled on
+so far in pipelines is a model where buffers are pushed by a producer (`IPipelineWriter`) to a consumer (`IPipelineReader`).
+
+```C#
+public interface IPipelineReader
+{
+    ReadableBufferAwaitable ReadAsync();
+    void Advance(ReadCursor consumed, ReadCursor examined);
+    void CancelPendingRead();
+    void Complete(Exception exception = null);
+}
+```
+
+```C#
+public interface IPipelineWriter
+{
+    Task Writing { get; }
+    WritableBuffer Alloc(int minimumSize = 0);
+    void Complete(Exception exception = null);
+}
+```
+
+The producer gets a handle on a `WritableBuffer` and the consumer gets a handle on a `ReadableBuffer` to do writes and reads respectively. 
+
+Memory in pipelines is managed as a linked list of `OwnedMemory<T>`:
+
+```C#
+public class BufferSegment
+{
+    // The backing memory, usually leased from a pool
+    OwnedMemory<byte> Data { get; set; }
+    public int Start { get; set; }
+    public int End { get; set; }
+    
+    // The next segment
+    BufferSegment Next { get; set; }
+    
+    // The segment cannot be written to
+    public bool ReadOnly { get; set; }
```

Comment:
> I don't know why we still seal things. 

1. Performance: easier for code gen technologies (mostly JIT) to analyze sealed types. Result is it can emit better code at runtime.
2. Design: an unsealed type with no virtual methods doesn't really serve much of a purpose (other than to make JIT harder).